firebase-firestore/src/main/java/com/google/firebase/firestore/remote/Datastore.java

```diff
@@ -109,9 +114,17 @@ public Datastore(
     // all calls need to be audited to make sure they are executed on the right thread.
     channelBuilder.executor(workerQueue.getExecutor());
 
+    // Wrap the ManagedChannelBuilder in an AndroidChannelBuilder. This allows the channel to
+    // respond more gracefully to network change events (such as switching from cell to wifi).
+    AndroidChannelBuilder androidChannelBuilder =
+        AndroidChannelBuilder.fromBuilder(channelBuilder).context(context);
```

Comment:
This may not be possible, but this seems like a lot of work to do in the constructor of Datastore.

Could we pass the completed FirestoreChannel or ManagedChannel as a constructor parameter instead?

This would allow us to use a non-android managed channel in unit tests but still do the full thing in production code.

Reply:
As discussed, lets defer.

Also as discussed: lets not submit to master, but otherwise, we can use this branch to troubleshoot issues with grpc reconnects. In the meantime, I'll create a PR to hookup shutdown logic to the channel. Once that lands, we can revisit and land this one too.

Reply:
SGTM.